File: javascript/153.find-minimum-in-rotated-sorted-array.js
/*
 * @lc app=leetcode id=153 lang=javascript
 *
 * [153] Find Minimum in Rotated Sorted Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
  // Special binary search
  // O(log(n)) time
  // O(k) space
  let low = 0
  let high = nums.length - 1
  let res = nums[0]
  while (low <= high) {
    if (nums[low] < nums[high]) {
      res = Math.min(res, nums[low])
      break
    }
    let midPoint = Math.floor((high + low) / 2)
    res = Math.min(res, nums[midPoint])
    if (nums[midPoint] >= nums[low]) {
      low = midPoint + 1
    } else {
      high = midPoint - 1
    }
  }
  return res
};
// @lc code=end

File: cpp/153.find-minimum-in-rotated-sorted-array.cpp
/*
 * @lc app=leetcode id=153 lang=cpp
 *
 * [153] Find Minimum in Rotated Sorted Array
 */

// @lc code=start
class Solution
{
public:
    int findMin(vector<int> &nums)
    {
        // brute force approach
        // iterate over nums, find min
        // O(n) time O(k) space
        // binary search approach
        // with two pointers
        // if midpoint is greater than indexH
        // than search mid-indexH
        // else serach indexL-mid
        // O(log(n)) time O(k) space
        int indexL = 0;
        int indexH = nums.size() - 1;
        int midPoint;
        while (indexL < indexH)
        {
            midPoint = (indexH + indexL) / 2;
            if (nums[midPoint] > nums[indexH])
            {
                indexL = midPoint + 1;
            }
            else
            {
                indexH = midPoint;
            }
        }
        return nums[indexL];
    }
};
// @lc code=end
